src/utils/ExchangeTokenRetrieval.tsx

```diff
@@ -0,0 +1,137 @@
+import { apolloClient as realApolloClient } from 'client/apolloClient'
+import { ExchangeTokenDocument } from 'data/graphql'
+import * as React from 'react'
+import { Mount } from 'react-lifecycle-components'
+import { useHistory } from 'react-router'
+import { captureSentryError } from 'utils/sentry-client'
+import {
+  StorageContainer,
+  StorageEffects,
+  StorageState,
+} from 'utils/StorageContainer'
+
+export enum ExchangeTokenRetrievalState {
+  Loading = 'Loading',
+  Success = 'Success',
+  Error = 'Error',
+  InvalidToken = 'InvalidToken',
+  ExpiredToken = 'ExpiredToken',
+  TakingTooLong = 'TakingTooLong',
+}
+
+export const ExchangeTokenRetrieval: React.FC<{
+  children: (props: {
+    exchangeTokenState: ExchangeTokenRetrievalState
+    retry: () => void
+  }) => React.ReactElement
+  apolloClient?: typeof realApolloClient
+}> = ({ children, apolloClient = realApolloClient }) => {
+  const [hasFinishedFakeLoading, setHasFinishedFakeLoading] = React.useState(
+    false,
+  )
+  const [actualLoadingState, setActualLoadingState] = React.useState(
+    ExchangeTokenRetrievalState.Loading,
+  )
+  const [tokenExchangeState, setTokenExchangeState] = React.useState(
+    ExchangeTokenRetrievalState.Loading,
+  )
+  const { location } = useHistory()
+
+  React.useEffect(() => {
+    if (!hasFinishedFakeLoading) {
+      return
+    }
+
+    setTokenExchangeState(actualLoadingState)
+  }, [actualLoadingState, hasFinishedFakeLoading])
+
+  React.useEffect(() => {
+    if (hasFinishedFakeLoading) {
+      return
+    }
+
+    setTimeout(() => {
+      setHasFinishedFakeLoading(true)
```

Comment:
Why do we need this fake loading state?

Reply:
It's too fast 😂 looks like a content flash if we don't debounce the loading state a bit imo.

Reply:
Ah, that's what I suspected 😄